### PR TITLE
feat: Added 'save' and 'load' state dictionary functionality

### PR DESCRIPTION
**Pull request by lsetiawan**
## Overview

This PR is jam packed with the new feature for a 'Simulator' to have a `state_dict` method that outputs an immutable state dictionary. Additionally, this state dictionary is able to be saved and loaded to a file with safetensors format. The reasoning to use safetensors format can be found in https://github.com/uw-ssec/caustics/issues/44.

### Person 1
```python
cosmology = caustics.cosmology.FlatLambdaCDM(name="cosmo", h0=None)

sie = caustics.lenses.SIE(cosmology=cosmology, name="sie", z_l=0.5, x0=0., y0=0., q = 0.4, phi = np.pi/2, b = 1.)
src = caustics.light.Sersic(name="src", x0=-0.2, y0=0., q=0.6, phi=-np.pi/4, n=1.5, Re=3., Ie= 1.)

sim = caustics.sims.Lens_Source(name="mysim", lens=sie, source=src, pixelscale=0.05, pixels_x=100)

# Save to a default state file in current working directory where code is run
fpath = sim.state_dict().save()
```

### Person 2
```python
# Setup similar objects first without knowing the parameters
cosmology = caustics.cosmology.FlatLambdaCDM(name="cosmo")

sie = caustics.lenses.SIE(cosmology=cosmology, name="sie")
src = caustics.light.Sersic(name="src")

sim = caustics.sims.Lens_Source(name="mysim", lens=sie, source=src, pixelscale=0.05, pixels_x=100)

# Load the state file from person 1 above
sim.load_state_dict(fpath)
```

Now person 2 would have sim object that has same parameters as person 1

### CODE REVIEW

File: tests/sims/conftest.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+from caustics.cosmology import FlatLambdaCDM
+
+
+@pytest.fixture
+def simple_common_sim():
+    class Sim(Simulator):
+        def __init__(self):
+            super().__init__()
+            self.cosmo = FlatLambdaCDM(h0=None)
+            self.epl = EPL(self.cosmo)
```
Comment (ConnorStoneAstro) [inline tests/sims/conftest.py:15]: could assign some values to the objects' parameter just for the sake of testing

**Review by ConnorStoneAstro — CHANGES_REQUESTED** [review]
This is very cool!

File: tests/helpers/sims.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+from caustics.namespace_dict import NestedNamespaceDict
+
+
+def extract_tensors(params, include_params=False):
```
[the remark this answers is not in the document]
Comment (lsetiawan) [reply tests/helpers/sims.py:4]: Right, thanks for pointing that out, I'll just extract that as a general utility function 😅

**Comment by lsetiawan** [comment]
Thanks @ConnorStoneAstro for your quick and thoughtful review. I've made the necessary adjustments from your suggestions. Please give it another go over 🙏🏽

**Review by ConnorStoneAstro — APPROVED** [review]
Looks good TM! :)

**Pull request merged**